Parent node is not automatically created in the Zookeeper discovery.

- [x] I have searched the [issues](https://github.com/seata/seata/issues) of this repository and believe that this is not a duplicate.

### Ⅰ. Issue Description
If zookeeper server is newly installed, It will throw exception when do register.
```
org.I0Itec.zkclient.exception.ZkNoNodeException: org.apache.zookeeper.KeeperException$NoNodeException: KeeperErrorCode = NoNode for /registry/zk/default/192.168.0.123:33333

	at org.I0Itec.zkclient.exception.ZkException.create(ZkException.java:47)
	at org.I0Itec.zkclient.ZkClient.retryUntilConnected(ZkClient.java:1001)
	at org.I0Itec.zkclient.ZkClient.create(ZkClient.java:528)
	at org.I0Itec.zkclient.ZkClient.create(ZkClient.java:499)
	at org.I0Itec.zkclient.ZkClient.createEphemeral(ZkClient.java:552)
	at io.seata.discovery.registry.zk.ZookeeperRegisterServiceImpl.doRegister(ZookeeperRegisterServiceImpl.java:102)
	at io.seata.discovery.registry.zk.ZookeeperRegisterServiceImpl.register(ZookeeperRegisterServiceImpl.java:94)
	at io.seata.discovery.registry.zk.ZookeeperRegisterServiceImplTest.testAll(ZookeeperRegisterServiceImplTest.java:78)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
Caused by: org.apache.zookeeper.KeeperException$NoNodeException: KeeperErrorCode = NoNode for /registry/zk/default/192.168.0.123:33333
	at org.apache.zookeeper.KeeperException.create(KeeperException.java:111)
	at org.apache.zookeeper.KeeperException.create(KeeperException.java:51)
	at org.apache.zookeeper.ZooKeeper.create(ZooKeeper.java:783)
	at org.I0Itec.zkclient.ZkConnection.create(ZkConnection.java:100)
	at org.I0Itec.zkclient.ZkClient$3.call(ZkClient.java:531)
	at org.I0Itec.zkclient.ZkClient$3.call(ZkClient.java:528)
	at org.I0Itec.zkclient.ZkClient.retryUntilConnected(ZkClient.java:991)
	... 30 more
```


### Ⅱ. Describe what happened

When register address `/registry/zk/default/192.168.0.123:33333` to zookeeper, and there is no parent path "/registry/zk/default/" exists.

### Ⅲ. Describe what you expected to happen

Auto creates the parent path if it is not present.

### Ⅵ. Environment:

- JDK version : 1.8.0
- OS : MacOSX
- Seata: 0.5.0

--- a/discovery/seata-discovery-zk/pom.xml
+++ b/discovery/seata-discovery-zk/pom.xml
@@ -43,11 +43,6 @@
         </dependency>
 
         <dependency>
-            <groupId>junit</groupId>
-            <artifactId>junit</artifactId>
-            <scope>test</scope>
-        </dependency>
-        <dependency>
             <groupId>org.apache.curator</groupId>
             <artifactId>curator-test</artifactId>
             <version>2.9.1</version>

--- a/discovery/seata-discovery-zk/pom.xml
+++ b/discovery/seata-discovery-zk/pom.xml
@@ -41,5 +41,17 @@
                 </exclusion>
             </exclusions>
         </dependency>
+
+        <dependency>
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.curator</groupId>
+            <artifactId>curator-test</artifactId>
+            <version>2.9.1</version>
+            <scope>test</scope>
+        </dependency>
     </dependencies>
 </project>

--- a/discovery/seata-discovery-zk/src/main/java/io/seata/discovery/registry/zk/ZookeeperRegisterServiceImpl.java
+++ b/discovery/seata-discovery-zk/src/main/java/io/seata/discovery/registry/zk/ZookeeperRegisterServiceImpl.java
@@ -98,12 +98,22 @@ public class ZookeeperRegisterServiceImpl implements RegistryService<IZkChildLis
         if (checkExists(path)) {
             return false;
         }
+        createParentIfNotPresent(path);
         getClientInstance().createEphemeral(path, true);
         REGISTERED_PATH_SET.add(path);
         return true;
     }
 
-
+    private void createParentIfNotPresent(String path) {
+        int i = path.lastIndexOf('/');
+        if (i > 0) {
+            String parent = path.substring(0, i);
+            if (!checkExists(parent)) {
+                getClientInstance().createPersistent(parent);
+            }
+        }
+    }
+    
     private boolean checkExists(String path) {
         return getClientInstance().exists(path);
     }
@@ -124,11 +134,12 @@ public class ZookeeperRegisterServiceImpl implements RegistryService<IZkChildLis
         }
 
         String path = ROOT_PATH + cluster;
-        if (getClientInstance().exists(path)) {
-            getClientInstance().subscribeChildChanges(path, listener);
-            LISTENER_SERVICE_MAP.putIfAbsent(cluster, new ArrayList<>());
-            LISTENER_SERVICE_MAP.get(cluster).add(listener);
+        if (!getClientInstance().exists(path)) {
+            getClientInstance().createPersistent(path);
         }
+        getClientInstance().subscribeChildChanges(path, listener);
+        LISTENER_SERVICE_MAP.putIfAbsent(cluster, new ArrayList<>());
+        LISTENER_SERVICE_MAP.get(cluster).add(listener);
     }
 
     @Override
@@ -167,7 +178,12 @@ public class ZookeeperRegisterServiceImpl implements RegistryService<IZkChildLis
             return null;
         }
 
-        Boolean exist = getClientInstance().exists(ROOT_PATH + clusterName);
+        return doLookup(clusterName);
+    }
+
+    // visible for test.
+    List<InetSocketAddress> doLookup(String clusterName) throws Exception {
+        boolean exist = getClientInstance().exists(ROOT_PATH + clusterName);
         if (!exist) {
             return null;
         }
@@ -188,30 +204,40 @@ public class ZookeeperRegisterServiceImpl implements RegistryService<IZkChildLis
 
     private ZkClient getClientInstance() {
         if (zkClient == null) {
-            zkClient = new ZkClient(FILE_CONFIG.getConfig(FILE_CONFIG_KEY_PREFIX + SERVER_ADDR_KEY),
-                FILE_CONFIG.getInt(FILE_CONFIG_KEY_PREFIX + SESSION_TIME_OUT_KEY),
-                FILE_CONFIG.getInt(FILE_CONFIG_KEY_PREFIX + CONNECT_TIME_OUT_KEY));
-            if (!zkClient.exists(ROOT_PATH_WITHOUT_SUFFIX)) {
-                zkClient.createPersistent(ROOT_PATH_WITHOUT_SUFFIX, true);
+            synchronized (ZookeeperRegisterServiceImpl.class) {
+                if (null == zkClient) {
+                    zkClient = buildZkClient(FILE_CONFIG.getConfig(FILE_CONFIG_KEY_PREFIX + SERVER_ADDR_KEY),
+                            FILE_CONFIG.getInt(FILE_CONFIG_KEY_PREFIX + SESSION_TIME_OUT_KEY),
+                            FILE_CONFIG.getInt(FILE_CONFIG_KEY_PREFIX + CONNECT_TIME_OUT_KEY));
+                }
             }
-            zkClient.subscribeStateChanges(new IZkStateListener() {
-
-                @Override
-                public void handleStateChanged(Watcher.Event.KeeperState keeperState) throws Exception {
-                    //ignore
-                }
-
-                @Override
-                public void handleNewSession() throws Exception {
-                    recover();
-                }
-
-                @Override
-                public void handleSessionEstablishmentError(Throwable throwable) throws Exception {
-                    //ignore
-                }
-            });
         }
+        return zkClient;
+    }
+    
+    // visible for test.
+    ZkClient buildZkClient(String address, int sessionTimeout, int connectTimeout) {
+        ZkClient zkClient = new ZkClient(address, sessionTimeout,connectTimeout);
+        if (!zkClient.exists(ROOT_PATH_WITHOUT_SUFFIX)) {
+            zkClient.createPersistent(ROOT_PATH_WITHOUT_SUFFIX, true);
+        }
+        zkClient.subscribeStateChanges(new IZkStateListener() {
+
+            @Override
+            public void handleStateChanged(Watcher.Event.KeeperState keeperState) throws Exception {
+                //ignore
+            }
+
+            @Override
+            public void handleNewSession() throws Exception {
+                recover();
+            }
+
+            @Override
+            public void handleSessionEstablishmentError(Throwable throwable) throws Exception {
+                //ignore
+            }
+        });
         return zkClient;
     }
 

--- a/discovery/seata-discovery-zk/src/test/java/io/seata/discovery/registry/zk/ZookeeperRegisterServiceImplTest.java
+++ b/discovery/seata-discovery-zk/src/test/java/io/seata/discovery/registry/zk/ZookeeperRegisterServiceImplTest.java
@@ -1,0 +1,109 @@
+/*
+ *  Copyright 1999-2019 Seata.io Group.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package io.seata.discovery.registry.zk;
+
+import io.seata.common.util.NetUtil;
+import io.seata.discovery.registry.RegistryService;
+import org.I0Itec.zkclient.IZkChildListener;
+import org.I0Itec.zkclient.ZkClient;
+import org.apache.curator.test.TestingServer;
+import org.junit.AfterClass;
+import org.junit.Assert;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * @author <a href="mailto:zhanggeng.zg@antfin.com">GengZhang</a>
+ */
+public class ZookeeperRegisterServiceImplTest {
+    protected static TestingServer server = null;
+    @BeforeClass
+    public static void adBeforeClass() {
+        try {
+            server = new TestingServer(2181, true);
+            server.start();
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+    }
+
+    @AfterClass
+    public static void adAfterClass() {
+        if (server != null) {
+            try {
+                server.stop();
+            } catch (IOException e) {
+                e.printStackTrace();
+            }
+        }
+    }
+
+    ZookeeperRegisterServiceImpl service  = (ZookeeperRegisterServiceImpl) new ZookeeperRegistryProvider().provide();
+    
+    @Test
+    public void getInstance() {
+        ZookeeperRegisterServiceImpl service1 = ZookeeperRegisterServiceImpl.getInstance();
+        Assert.assertEquals(service1, service);
+    }
+    
+    @Test
+    public void buildZkTest() {
+        ZkClient client = service.buildZkClient("127.0.0.1:2181", 5000, 5000);
+        Assert.assertTrue(client.exists("/zookeeper"));
+    }
+
+    @Test
+    public void testAll() throws Exception {
+        service.register(new InetSocketAddress(NetUtil.getLocalAddress(), 33333));
+
+        Assert.assertNull(service.lookup("xxx"));
+        List<InetSocketAddress> lookup2 = service.doLookup("default");
+        Assert.assertTrue(lookup2.size() == 1);
+        
+        final List<String> data = new ArrayList<>();
+        final CountDownLatch latch = new CountDownLatch(1);
+        IZkChildListener listener = (s, list) -> {
+            data.clear();
+            data.addAll(list);
+            latch.countDown();
+        };
+        service.subscribe("default", listener);
+        final CountDownLatch latch2 = new CountDownLatch(1);
+        final List<String> data2 = new ArrayList<>();
+        IZkChildListener listener2 = (s, list) -> {
+            data2.clear();
+            data2.addAll(list);
+            latch2.countDown();
+        };
+        service.subscribe("default", listener2);
+        
+        service.unregister(new InetSocketAddress(NetUtil.getLocalAddress(), 33333));
+        latch2.await(1000, TimeUnit.MILLISECONDS);
+        Assert.assertTrue(data2.size() == 0);
+        
+        service.unsubscribe("default", listener);
+        service.unsubscribe("default", listener2);
+    }
+
+}

--- a/discovery/seata-discovery-zk/src/test/java/io/seata/discovery/registry/zk/ZookeeperRegisterServiceImplTest.java
+++ b/discovery/seata-discovery-zk/src/test/java/io/seata/discovery/registry/zk/ZookeeperRegisterServiceImplTest.java
@@ -16,7 +16,6 @@
 package io.seata.discovery.registry.zk;
 
 import io.seata.common.util.NetUtil;
-import io.seata.discovery.registry.RegistryService;
 import org.I0Itec.zkclient.IZkChildListener;
 import org.I0Itec.zkclient.ZkClient;
 import org.apache.curator.test.TestingServer;
@@ -25,8 +24,6 @@ import org.junit.Assert;
 import org.junit.BeforeClass;
 import org.junit.Test;
 
-import java.io.IOException;
-import java.net.InetAddress;
 import java.net.InetSocketAddress;
 import java.util.ArrayList;
 import java.util.List;
@@ -34,28 +31,20 @@ import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.TimeUnit;
 
 /**
- * @author <a href="mailto:zhanggeng.zg@antfin.com">GengZhang</a>
+ * @author Geng Zhang
  */
 public class ZookeeperRegisterServiceImplTest {
     protected static TestingServer server = null;
     @BeforeClass
-    public static void adBeforeClass() {
-        try {
-            server = new TestingServer(2181, true);
-            server.start();
-        } catch (Exception e) {
-            e.printStackTrace();
-        }
+    public static void adBeforeClass() throws Exception {
+        server = new TestingServer(2181, true);
+        server.start();
     }
 
     @AfterClass
-    public static void adAfterClass() {
+    public static void adAfterClass() throws Exception {
         if (server != null) {
-            try {
-                server.stop();
-            } catch (IOException e) {
-                e.printStackTrace();
-            }
+            server.stop();
         }
     }
 

--- a/discovery/seata-discovery-zk/src/test/java/io/seata/discovery/registry/zk/ZookeeperRegistryProviderTest.java
+++ b/discovery/seata-discovery-zk/src/test/java/io/seata/discovery/registry/zk/ZookeeperRegistryProviderTest.java
@@ -20,7 +20,7 @@ import org.junit.Assert;
 import org.junit.Test;
 
 /**
- * @author <a href="mailto:zhanggeng.zg@antfin.com">GengZhang</a>
+ * @author Geng Zhang
  */
 public class ZookeeperRegistryProviderTest {
 
@@ -28,6 +28,7 @@ public class ZookeeperRegistryProviderTest {
     public void provide() {
         ZookeeperRegistryProvider provider = new ZookeeperRegistryProvider();
         RegistryService registryService = provider.provide();
+        Assert.assertTrue(registryService instanceof ZookeeperRegisterServiceImpl);
         Assert.assertEquals(registryService, provider.provide());
     }
 

--- a/discovery/seata-discovery-zk/src/test/java/io/seata/discovery/registry/zk/ZookeeperRegistryProviderTest.java
+++ b/discovery/seata-discovery-zk/src/test/java/io/seata/discovery/registry/zk/ZookeeperRegistryProviderTest.java
@@ -1,0 +1,34 @@
+/*
+ *  Copyright 1999-2019 Seata.io Group.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package io.seata.discovery.registry.zk;
+
+import io.seata.discovery.registry.RegistryService;
+import org.junit.Assert;
+import org.junit.Test;
+
+/**
+ * @author <a href="mailto:zhanggeng.zg@antfin.com">GengZhang</a>
+ */
+public class ZookeeperRegistryProviderTest {
+
+    @Test
+    public void provide() {
+        ZookeeperRegistryProvider provider = new ZookeeperRegistryProvider();
+        RegistryService registryService = provider.provide();
+        Assert.assertEquals(registryService, provider.provide());
+    }
+
+}

--- a/discovery/seata-discovery-zk/src/test/java/io/seata/discovery/registry/zk/ZookeeperRegistryProviderTest.java
+++ b/discovery/seata-discovery-zk/src/test/java/io/seata/discovery/registry/zk/ZookeeperRegistryProviderTest.java
@@ -15,7 +15,6 @@
  */
 package io.seata.discovery.registry.zk;
 
-import io.seata.discovery.registry.RegistryService;
 import org.junit.Assert;
 import org.junit.Test;
 

--- a/discovery/seata-discovery-zk/src/test/java/io/seata/discovery/registry/zk/ZookeeperRegistryProviderTest.java
+++ b/discovery/seata-discovery-zk/src/test/java/io/seata/discovery/registry/zk/ZookeeperRegistryProviderTest.java
@@ -27,9 +27,7 @@ public class ZookeeperRegistryProviderTest {
     @Test
     public void provide() {
         ZookeeperRegistryProvider provider = new ZookeeperRegistryProvider();
-        RegistryService registryService = provider.provide();
-        Assert.assertTrue(registryService instanceof ZookeeperRegisterServiceImpl);
-        Assert.assertEquals(registryService, provider.provide());
+        Assert.assertTrue(provider.provide() instanceof ZookeeperRegisterServiceImpl);
     }
 
 }

